### PR TITLE
refactor: use hooks to extend bulk_transaction

### PR DESCRIPTION
**Pull request by kittiu**
Bulk Transaction is very useful feature, but now it is really difficult to add more methods.

When creating a new app, there is a chance that the the new kind of bulk transaction is required. 

This PR move the mapper function inside a Doctype so it can be extened by super() in other app (I can't find a proper doctype other than Bulk Transaction Detail Log to place the get_mapper()).

I.e., I have a new app, and so I can do something like this with override_doctype_class.

```
from erpnext.bulk_transaction.doctype.bulk_transaction_log_detail.bulk_transaction_log_detail import BulkTransactionLogDetail


class BulkTransactionLogDetailThaiPayroll(BulkTransactionLogDetail):
    
    def get_mapper(self):
        from thai_payroll.custom import salary_slip
        mapper = super().get_mapper()
        mapper.update({
            "Salary Slip": {
		    	"Withholding Tax Cert Employee": salary_slip.make_withholding_tax_cert_employee,
		    },
        })
        return mapper
```

Note: IMHO, writing method within a class should be a preferred way to allow easier extension. Currently, in many case, I always need to do monkey patch and method overwrite (which is not good for upgrade).

### CODE REVIEW

**Comment by blaggacao** [comment]
I believe the predominant pattern for downstream extensions of upstream behaviour is to declare hooks. While I personally also prefer inheritance, in this case I agree that the overall frappe-style would be to define a hook in hooks.py and load that correspondingly.

**Comment by kittiu** [comment]
> I believe the predominant pattern for downstream extensions of upstream behaviour is to declare hooks. While I personally also prefer inheritance, in this case I agree that the overall frappe-style would be to define a hook in hooks.py and load that correspondingly.

Thanks for the advice. Me too from Odoo world and not yet familiar frappe style. In this case, how do you think we should do to make it hookable? Any suggestion and/or code example? I am keen to adapt

**Comment by kittiu** [comment]
A question though. When you say to create a new hook on upstream. In this case, it is on erpnext app not in the framework.
Is it advisable to create additional hook outside of the frappe framework?

**Comment by blaggacao** [comment]
I'm on the road, so I'll be brief / sparse:

In Frappe:

- There's a function (forgot the name) to retrieve hooks, say `frappe.get_hooks` (?)
- Revise how that method is implemented (grep your way around, if need be, or use an lsp to jump around)

In ERPNext:

- Add that method calling your newly named hook to the bulk code to call and expand the mapping for new documents

Note, how hooks aren't declared anywhere, just queried.

Downstream:

- Use that hook in a downstream project and test it and notify the test results here

That's how I would approach it. Let me know if you have any more questions, I'll try to be helpful.

**Comment by kittiu** [comment]
In this case, so you think, the Bulk Transaction featurs/doctypes should be moving into Frappe instead of ERPNext.

And so, ERPNext is probably first downstream app, and in its hook.py could be,

bulk_transaction = [
   'erpnext.utils.bulk_transaction.task'
]

Do I understand your explaination correctly?

**Comment by blaggacao** [comment]
> In this case, so you think, the Bulk Transaction featurs/doctypes should be moving into Frappe instead of ERPNext.

That's not necessary or implied. You can (should) leave the implementation of bulk transactions inside ERPNext, but you can _find_ the hook implementation in Frappe.

> ERPNext is probably first downstream app

Correct. For my part, I'll leave it to you for now to play with a concrete implementation. Next touch point an updated & tested PR? :handshake: 

PS: I'm not a maintainer and don't have special authority over the decision to merge, just trying to guide in a direction that I feel has better chances of getting accepted.

**Comment by kittiu** [comment]
Okay I think I get your point. Will be back after I have cooked something.

File: erpnext/utilities/bulk_transaction.py
Comment (kittiu) [inline erpnext/utilities/bulk_transaction.py]: This part is being used to get the mapper function from hooks.py

Comment (kittiu) [reply erpnext/utilities/bulk_transaction.py]: @blaggacao I hope I am not wrong. Thanks for your hint.
And it now make more sense to me about Frappe ways.

Comment (kittiu) [reply erpnext/utilities/bulk_transaction.py]: With this, the downstream can do something like this,
https://github.com/kittiu/thai_payroll/pull/7/files

Comment (blaggacao) [reply erpnext/utilities/bulk_transaction.py]: I would advise against dogfooding the mapper within ERPNext itself.

While this is generally a good idea, consider the downside: now, there's one non-lsp-transparent indirection in the code path, meanig you can't just lsp-jump around (e.g. on GitHub) anymore, so overall the code becomes harder to read and maintain.

We can avoid this by just putting the hook snippet in place of the orginal code exclusively for eventual downstream use.

It's also a smaller, more trivial change, which is a plus for review.

Comment (blaggacao) [reply erpnext/utilities/bulk_transaction.py]: PS: your downstream code looks much cleaner, now :+1:

Comment (kittiu) [reply erpnext/utilities/bulk_transaction.py]: > I would advise against dogfooding the mapper within ERPNext itself.
> 
> While this is generally a good idea, consider the downside: now, there's one non-lsp-transparent indirection in the code path, meanig you can't just lsp-jump around (e.g. on GitHub) anymore, so overall the code becomes harder to read and maintain.
> 
> We can avoid this by just putting the hook snippet in place of the orginal code exclusively for eventual downstream use.
> 
> It's also a smaller, more trivial change, which is a plus for review.

You mean not to change the existing, just provide the hook point?
I just worry that the downstream coder won't know what to do. No sample code.

Comment (blaggacao) [reply erpnext/utilities/bulk_transaction.py]: Yeah, got ya. This is true in principle, but I think maintainability primes over education.

We'd probably other tools to educate. For example consistent LSP support for all things. That would serve the same goal, but better.

**Comment by kittiu** [comment]
@blaggacao thank you for your review. So I would call this a done. If there is anything else for this to get accepted?

**Review by blaggacao — CHANGES_REQUESTED** [review]
I have a preference for _not_ dogfooding the hook here on the basis that using the hook here breaks lsp-jumps (the functionality that an LSP through static code analysis can identify the definition and calling sites of an identifier).

I always found it extremely annoying in the codebase, when a hook was used when another option existed, because it makes it very hard to trace the code and very easy to miss something in the business logic during audit.

This is also true in principle and practice here. Hence, I request hereby to change this PR to _not_ touch `hooks.py`, at all.

I have considered the value of having an example in tree, but in my opinion, the breakage of static code analysis weighs much more heavily as Frappe tries to move to a ever more type-hinted codebase.

<sub>of course this argument can be overruled by a <i>maintainer</i></sub>

**Comment by kittiu** [comment]
@blaggacao here you go. Just provide the hooks point.

**Review by blaggacao — APPROVED** [review]

**Comment by blaggacao** [comment]
@ruthra-kumar this seems a sufficiently small & useful change. The author already uses it downstream in localization, my motion would be to merge.

**Comment by ruthra-kumar** [comment]
@blaggacao 
Awesome work on reviewing and providing sensible suggestions on the structure of the change.

**Comment by ruthra-kumar** [comment]
@kittiu 
Allowing hooks to make bulk processing extensible for downstream apps makes sense. Thanks for your contribution.

**Review by ruthra-kumar — APPROVED** [review]

**Comment by kittiu** [comment]
Thanks @blaggacao @ruthra-kumar

**Pull request merged**
